### ISSUE TITLE
Trying to make OCaml bindings; Cannot build because of missing headers

### ISSUE THREAD

**Issue by UberPyro** [fadoss/maude-bindings#5]
Hello, I am trying to use the code here to make OCaml bindings for Maude. However, I'm running into some strange issues trying to generate the bindings. I'm on Ubuntu 20.04 using the modified Maude 3.3.1 and SWIG 4.0.1; additional info from cmake [here](https://gist.github.com/UberPyro/94c095c67c418e41ba523134a21f4c6c). 

Running the `git submodule update --init` command right away does not work because the `maudesmc` folder exists and is not empty. After deleting its contents, the command clones the repository. 

Following this, trying to run the python setup file does not work. [It fails with this error.](https://gist.github.com/UberPyro/3758fed3b555e48b23bb495dc0314143)

No matter, we can install Maude manually. 

Following the releases link, installing the anylinux release provides us the config file, an so file, and the maude and simaude executables. Presumably the config.h file goes into the build directory and the so file goes into the lib directory. 

After doing this, I create an `ocaml-build` folder in the soure tree, and run
```
$ cmake .. -DLANGUAGE=ocaml -DBUILD_LIBMAUDE=OFF
$ cmake --build .
```
This gives an error about a missing `caml/alloc.h` in an include, which is somewhat expected. 

Providing SWIG the path to my ocaml runtime by editting `CMakeLists.txt` seems to resolve this. At the end of the language-specific if-chain: 
```
elseif (${LANGUAGE} STREQUAL ocaml)
	list(APPEND EXTRA_INCLUDE_DIRS " /home/pyro/.opam/default/lib/ocaml ")
```
This resolves the ocaml-specific issue, but now there remain many maude-specific issues. 

Now I receive errors such as this: 
```
/home/pyro/old-maude/b/maude-bindings/ocaml-build/CMakeFiles/_maude.dir/maudeOCAML_wrap.cxx:1820:10: fatal error: macros.hh: No such file or directory
 1820 | #include "macros.hh"
      |          ^~~~~~~~~~~
```
[Full error here. ](https://gist.github.com/UberPyro/e309c18a4cd72690f9eaf753bbf93439)

I believe these are the cmake header files associated with Maude. I find this confusing as I would expect these to already be included; I wouldn't have to do this myself. I assume there is something wrong at this point with what I have done? 

In any case, we can resolve these issues by continuing to add every folder in `Maude/src`, e.g. `list(APPEND EXTRA_INCLUDE_DIRS " /home/pyro/Maude/src/Utility ")` and so on to `CMakeLists.txt`. We also have to add the header file `bdd.h` from `buddy/src` as well. 

This now produces an error about a missing `specialHubSymbol.hh`. This one is especially strange, as it's a cmake header file that doesn't seem to be in the Maude repository. We can disable the code dealing with hooks from `maude_wrappers.hh`. 

Following this, we receive an error about a `strategyTransitionGraph.hh` not being found. The code here cannot be easily disabled. 

Basically I am wondering: do you know what it is that I am missing that is causing these header files to not be found, and how I can fix it? None of the paths I am aware of here seem to be allowing me to build the ocaml bindings. Let me know if there's any more information I could provide. Thank you very much in advance.

**Comment by ningit** [comment]
Hello, I cannot reproduce the first part of the issue. The following sequence of commands runs without trouble for me:
```
$ git clone https://github.com/fadoss/maude-bindings
$ cd maude-bindings
$ git submodule update --init
$ python setup.py bdist_wheel
```
The `subprojects/maudesmc` folder exists after cloning the repository, but it is empty as it should. Then, `git submodule update --init` works and clones the `specials` branch of the modified Maude [repository](https://github.com/fadoss/maudesmc/tree/specials) (the [submodule](https://github.com/fadoss/maude-bindings/tree/master/subprojects) is linked to that commit). The missing `build` directory of the `setup-error.txt` exists in this branch, and the files  `specialHubSymbol.hh` and `strategyTransitionGraph.hh` are actually there. Finally, the `python setup.py` invocation succeeds (building Maude from source) and generates the Python package.

It seems that you are cloning a different branch of the [fadoss/maudesmc](https://github.com/fadoss/maudesmc) repository in the submodule, but this should not happen because the submodule is linked to a specific commit.

I have not tried to build the bindings for OCaml yet, but there should be no need to change the include directories. You are right with where to copy the binaries with the `-DBUILD_LIBMAUDE=OFF` option.

**Comment by UberPyro** [comment]
Thank you for the clear instructions on what to run. Everything works until I run `python setup.py bdist_wheel`, which exits with error `meson.build:147:0: ERROR: C++ library 'bdd' not found`. However, I followed the instructions for BuDDy and should have it installed, my `/usr/local/include` contains `bdd.h` and `/usr/local/bin` contains the appropriate libbdd `.a`, `la`, `.so` files and so on. Do you know if the build system is looking for BuDDy in a particular place? Thanks. [Full log. ](https://gist.github.com/UberPyro/da377e8864e47c886da0d5bbf5178864)

Edit: this is definitely an issue with pkg-config not finding bdd, so not an issue with maude-bindings, disregard.

**Comment by UberPyro** [comment]
With regards to the second error, I realized I had to `git submodule update --init` first, then manually create the `/installdir/lib` and `/build` folders with the appropriate header and library files in them, which eliminates the header errors but nets me a very complicated error. Would you happen to know what is going wrong here? 
[log.txt](https://github.com/fadoss/maude-bindings/files/11852431/log.txt)

**Comment by UberPyro** [comment]
The above errors were due to using OCaml 5 instead of OCaml 4, which SWIG didn't support. I think I'm on track for getting the OCaml bindings to work, thank you again for your help.

**Comment by ningit** [comment]
Sorry for the delayed answer.

> Everything works until I run `python setup.py bdist_wheel`, which exits with error `meson.build:147:0: ERROR: C++ library 'bdd' not found`. [...] Edit: this is definitely an issue with pkg-config not finding bdd, so not an issue with maude-bindings, disregard.

The distribution of BuDDy and its Ubuntu package do not come with a pkg-config file, and so the Meson build script does not use this method to find it. It simply looks for the `.so` and `.h` files. It seems that Meson does not search by default in `/usr/local`, but you can run `meson configure -Dextra-include-dirs=/usr/local/include -Dextra-lib-dirs=/usr/local/lib` in the subproject to add these paths.

Another option is skipping the compilation of Maude with the `-DBUILD_LIBMAUDE=OFF`.

> a very complicated error. Would you happen to know what is going wrong here?

In addition to the unsupported OCaml version, there are apparently other issues with the C++ STL.

**Comment by UberPyro** [comment]
Hi again, 

I was able to build the OCaml bindings for Maude, and was able to link the shared object file against OCaml programs. It looks like it is working, though I haven't been able to test any significant programs yet. 

Unfortunately the generated bindings are very opaque - everything is in terms of "c_obj", which is a sum type over C types, and it's up to the programmer to know what type to provide everywhere. I was able to manage with this, but I can't figure out how to produce a "Vector" type, specifically as I am trying to use the unify command. It seems that "Vector" is an object defined by Maude and not by SWIG alone. I have some helper functions, one is `array_to_vector : 'a c_obj_t -> ('b -> 'a c_obj_t) -> 'b array -> 'a c_obj_t`, but I'm quite lost as to what to do. Would you happen to know anything about how Maude vectors are encoded in C types? Thanks again. 

Edit: Nevermind, I have everything figured now, thanks again for everything.

**Comment by ningit** [comment]
Congratulations. It seems quite hard to make them work for OCaml.

The Javadoc [documentation](https://fadoss.github.io/maude-bindings/javadoc/es/ucm/maude/bindings/TermVector.html) may help you find the appropriate types for the functions. From the contents of the generated wrapper, it seems that there is a `new_TermVector` that takes a list of arguments, and produces an empty term vector given an empty list. Then, there is apparently a `TermVector_push_back` method/function to add elements. I suppose you have already found that.

The experimental OCaml generator in Swig produces this kind of opaque modules (as explained in the [documentation](https://www.swig.org/Doc4.1/Ocaml.html)).

**Comment by UberPyro** [comment]
Sorry to keep annoying you, just 1 more question. Is there any way to get the name of a symbol? You can find symbols using their name and other data, but there doesn't seem to be [any way to just get the name of a symbol](https://fadoss.github.io/maude-bindings/javadoc/es/ucm/maude/bindings/Symbol.html). Do you know how I could do this? I don't think it's the toString() method and I don't seem to have that in my generated bindings anyway. Thanks.

**Comment by ningit** [comment]
In Java, it is the `toString` method (`__repr__` in Python, `ToString` in C#, etc). These names are defined in [`common.i`](https://github.com/fadoss/maude-bindings/blob/765b8dbc8b8f1de491d8a1c74aef2895e40b92e7/swig/common.i#L18-L30), but there is no default name for unsupported languages. Hence, this function is not generated for OCaml.

I should add a default name for unsupported languages in the `#else` branch. Is there a standard name for this in OCaml?

**Comment by UberPyro** [comment]
Yes, `string_of_{datatype}`. Thanks. 

edit: `to_string` is fine too if that fits the pattern better.

**Comment by ningit** [comment]
I have committed a change so that a `to_string` method is generated for Ocaml (it fits the pattern better).

**Comment by UberPyro** [comment]
Thanks! Another thing I was wondering is if there was some way to produce variables as a term? There doesn't seem to be a way to create something like `X:Monoid` (assuming I defined the sort Monoid) using the API. I'm trying to build up terms with variables for a unification problem.

**Comment by ningit** [comment]
The [`parseTerm`](https://fadoss.github.io/maude-bindings/#maude.Module.parseTerm) method of the desired `Module` object with a string like `{name}:{sort}`. There is no specific method to build variables.
